installinator/src/reporter.rs

```diff
@@ -49,129 +78,418 @@ impl ProgressReporter {
-            // deterministically get the update. Need to decide post-PVT1.
-            let last_reported = report.last_seen;
-            let results: Vec<_> =
-                futures::stream::iter(peers.peers().iter().copied())
-                    .map(|peer| {
-                        report_backend.send_report_to_peer(
-                            peer,
-                            update_id,
-                            report.clone(),
-                        )
-                    })
-                    .buffer_unordered(8)
-                    .collect()
-                    .await;
-
-            if results.iter().any(|res| res.is_ok()) {
-                Some(last_reported)
-            } else {
-                None
+                // For each peer in diff.added, spawn a new report task.
+                for peer in diff.added {
+                    let task = ReportTask::new(
+                        *peer,
+                        update_id,
+                        // Note: report_rx.clone() starts off as seen. However,
+                        // the initial state of report_rx is None, so if there's
+                        // an report in the receiver, it will always be treated
+                        // as unseen. This means that if a report is available,
+                        // the task will immediately dispatch it to the peer.
+                        report_rx.clone(),
+                        backend.clone(),
+                    );
+                    peer_tasks.insert(*peer, task);
+                }
             }
         })
     }
+}
 
-    async fn on_tick(&mut self) {
-        if let Some(task) = self.on_tick_task.take() {
-            // If the task we spawned on a previous tick is still running, just
-            // put it back; we'll wait for the next tick.
-            if !task.is_finished() {
-                self.on_tick_task = Some(task);
+async fn discover_task_loop(
+    log: &slog::Logger,
+    report_backend: ReportProgressBackend,
+    peers_tx: watch::Sender<Option<PeerAddresses>>,
+) {
+    let mut discover_interval = time::interval(Duration::from_secs(5));
+    // MissedTickBehavior::Skip ensures that discovery happens every
+    // interval, rather than bursting in case report_backend.discover_peers
+    // takes a long time.
+    discover_interval.set_missed_tick_behavior(time::MissedTickBehavior::Skip);
+    // Do not tick discover_interval immediately. We'd like one discovery to
+    // happen at the start of the loop.
+
+    loop {
+        discover_interval.tick().await;
+        let peers = match report_backend.discover_peers().await {
+            Ok(peers) => peers,
+            Err(DiscoverPeersError::Retry(error)) => {
+                slog::warn!(
+                    log,
+                    "failed to discover peers, will retry: {}",
+                    DisplayErrorChain::new(error.as_error()),
+                );
                 return;
```

Comment:
Should this be `continue`?

Reply:
ack -- it should, whoops! I really should add tests for mock peer network flakiness, huh

Reply:
added some basic flakiness simulation which caught this bug right away. thanks for your eagle eyes as usual!